### ISSUE TITLE
Type '{ className: string; }' is not assignable to type 'IntrinsicAttributes & Props'. when using generics in component props. 

### ISSUE THREAD

**Issue by rostislav-simonik-plc** [microsoft/TypeScript#40383]
**TypeScript Version:**  3.9.7

**Search Terms:**
Type  is not assignable to type 'IntrinsicAttributes & Props'.
Property does not exist on type 'IntrinsicAttributes & Props'

**Code**

This doesn't work. 
```ts
type Props = {
  some?: string
}

export const SomeContainer = <PROPS extends Props, >({ some, ...rest }: PROPS): JSX.Element => {
  console.log(some)
  return (
    <div { ...rest} />
  )
}

export const a = (
  <SomeContainer
    className='someClassName'
    // some='a' // <<<< HERE IS THE CHANGE
  />
)
```

This works
```ts
type Props = {
  some?: string
}

export const SomeContainer = <PROPS extends Props, >({ some, ...rest }: PROPS): JSX.Element => {
  console.log(some)
  return (
    <div { ...rest} />
  )
}

export const a = (
  <SomeContainer
    className='someClassName'
    some='a'  // <<<< HERE IS THE CHANGE
  />
)
```

**Expected behavior:**
Both cases should not throw an error. 

**Actual behavior:**
The first one throws:

```
JSX attribute) className: string
Type '{ className: string; }' is not assignable to type 'IntrinsicAttributes & Props'.
  Property 'className' does not exist on type 'IntrinsicAttributes & Props'.ts(2322)
```

**Comment by RyanCavanaugh** [comment]
We will have someone look into this

**Comment by weswigham** [comment]
So, I looked at this for longer than I should have needed to before 🤦 ing and realizing nothing was wrong - `className` isn't a valid prop to pass because it's not mentioned in the constraint, and with no other valid props passed, gets flagged as excess (since the props being supplied share nothing in common with the weak generic constraint). `className` isn't a valid prop to pass because the only intrinsic property name is `key` (and in some versions `ref`) - other props, like `className`, are specific to _HTML_ elements. If you wanna explicitly take HTML attributes (because you're forwarding to an HTML tag) you want to include `React.HTMLAttributes` in your constraint - eg, [like so](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQMAnmFnAArFjoC8cA3nThxUJLAH4AXMJhRgAOwDmdAL506WAB6RYcXBDmp4AZVEBhfTGTysUOLwA8bDAHk2RuJphY5AE3QcILjgAMkwqGAA6AAkAFQBZABkAQRgZYAAjAFcvVHtkOUYAPgAaQoAKPmFRYrgIuspDOGUpJ1cjAEopACkjAA0IgFEAGywQb3huQv5BXX0REYihiAUykTH2mcoYTKg5ODKZoXsfYAA3flr6rENlOAB6QpmN1XUtaHg9A3hkO32Z+xMY3Mcks1igh10QzQqAAcsgxtwAORrLCmKGoWHwrCIiF3O5VBGI5CI+74+zk+xwKIDDADOAASXcMWpcFMUSSMIA4gMZg86O0gA).

**Comment by rostislav-simonik** [comment]
@weswigham  thanks for answer, I understand, i'll do that way,  but still wondering why such warning is not present when prop `some` is present. Shouldn't be this unified somehow?

**Comment by weswigham** [comment]
We call types with exclusively optional properties "weak types", because they do very little to constrain a type, and consider assignments to them which have _zero_ common properties between the source and target as likely an error (even though under other rules they would not be). Once the `some` property is present, that overlap exists, and we assume you know what you're doing with the other properties.

**Comment by nvmnghia** [comment]
Hi, is there any guide on TypeScript typeset for props?

Currently I do as follow:

```typescript
interface MyCompProps {
	someAttr: number
}
```

I then want to use `className` and `aria`, but listing them manually is incredibly cumbersome. I look around and found nothing until I hit this issue. Now it's clear for me: `interface MyCompProps extends React.HTMLAttributes<any>`, though I still don't fully understand it as I don't have an extensive guide on this matter.

**Comment by Inna-Mykytiuk** [comment]
Help me please. I have this issue
Type '{ setEditTodo: Dispatch<SetStateAction<string | null>>; }' is not assignable to type 'IntrinsicAttributes & ToDoState'.
  Property 'setEditTodo' does not exist on type 'IntrinsicAttributes & ToDoState'.
  
"use client";

import React, { useState } from "react";
import { BsEmojiSmile } from "react-icons/bs";
import { AiOutlinePlus } from "react-icons/ai";
import Picker from "@emoji-mart/react";
import data from "@emoji-mart/data";
import { TodoItem } from "./TodoItem";
import { useAppSelector, useAppDispatch } from "../../redux/hooks";
import { addTodo } from "../../redux/slice/todoSlice";

// interface TodoSetProps {
//   setEditTodo: React.Dispatch<React.SetStateAction<string | null>>;
// }

interface TodoItemProps {
  id: number;
  text: string;
  date: Date;
  completed: boolean;
  setEditTodo: React.Dispatch<React.SetStateAction<string | null>>;
}
interface ToDoState {
  todoList: TodoItemProps[];
}

const Todo: React.FC<ToDoState> = () => {
  const [showEmoji, setShowEmoji] = useState<boolean>(false);
  const [text, setText] = useState<string>("");
  const [editTodo, setEditTodo] = useState<string | null>(null);

  const dispatch = useAppDispatch();

  // add emoji
  const addEmoji = (e: any) => {
    const sym = e.unified.split("_");
    const codeArray: any = [];
    sym.forEach((el: any) => codeArray.push("0x" + el));
    let emoji = String.fromCodePoint(...codeArray);
    setText(text + emoji);
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    dispatch(addTodo({
      id: Math.floor(Math.random() * 100000),
      text,
      date: new Date().toISOString(),
      completed: false,
    }))


  }

  return (
    <div className="pt-3rem w-[90%] sm:w-[70%] md:w-[60%] lg:w-[40%] mx-auto mt-[11rem]">
      <h1 className="text-2xl font-medium text-center capitalize pb-[2rem]">
        tailwind todo list
      </h1>

      {/* todo input  */}
      <form
        onSubmit={handleSubmit}
        className="flex items-start gap-2 pt-2rem">
        <div className="w-full flex items-end p-2 bg-todo rounded-sm relative">
          <textarea
            value={text}
            onChange={(e) => setText(e.target.value)}
            placeholder="write your text"
            className="w-full bg-transparent outline-none resize-none text-sm"
            cols={30}
            rows={2}></textarea>

          <span
            onClick={() => setShowEmoji(!showEmoji)}
            className="cursor-pointer hover:text-slate-300">
            <BsEmojiSmile />
          </span>

          {showEmoji && (
            <div className="absolute top-[100%] right-2">
              <Picker
                data={data}
                emojiSize={20}
                emojiButtonSize={28}
                onEmojiSelect={addEmoji}
                maxFrequentRows={0}
              />
            </div>
          )}
        </div>

        <button
          className="flex items-center capitalize gap-2 bg-yellow-200 text-black py-1.5
            px-3 rounded-sm hover:bg-yellow-100">
          <AiOutlinePlus />
          {editTodo ? "update" : "add"}
        </button>
      </form>

      {/* print the todo lists  */}
      <div className="mt-[2rem]">
        <TodoItem setEditTodo={setEditTodo} />
      </div>
    </div>
  );
};

export { Todo }

"use client";
import {
  AiTwotoneDelete,
  AiTwotoneEdit,
  AiOutlineCheckCircle,
} from "react-icons/ai";
import { FaTimes } from "react-icons/fa";

import React, { useState } from "react";
import moment from "moment/moment";
import Layer from "./Layer";

// interface TodoSetProps {
//   setEditTodo: React.Dispatch<React.SetStateAction<string | null>>;
// }

interface TodoItemProps {
  id: number;
  text: string;
  date: Date;
  completed: boolean;
  setEditTodo: React.Dispatch<React.SetStateAction<string | null>>;
}
interface ToDoState {
  todoList: TodoItemProps[];
}

const TodoItem: React.FC<ToDoState> = () => {
  const [showText, setShowText] = useState(false);
  const [showFullText, setShowFullText] = useState("");

  const fullText = (id: number) => {
    const todoText = todo.find((todo) => todo.id === id);
    if (todoText) {
      setShowFullText(todoText.text);
      setShowText(true);
    }
  };

  const todo = [
    {
      id: 1,
      text: "Hello world",
      date: new Date(),
      completed: false,
    },
    {
      id: 2,
      text: "Hello world",
      date: new Date(),
      completed: false,
    },
  ];

  return (
    <div id="grid">
      {todo.map((todo, i) => (
        <div key={i} className="bg-todo p-2 rounded-md w-full h-full">
          <span className="text-xs text-slate-400">
            {moment(todo.date).fromNow()}
          </span>

          <div className="flex flex-col justify-between h-[80%]">
            <h1
              className={`pt-3 text-sm ${todo.completed ? "line-through text-slate-400" : ""
                }`}>
              {todo.text.substring(0, 36)}
              {todo.text.length > 36 && (
                <button
                  onClick={() => fullText(todo.id)}
                  className="text-red-400 text-xs hover:text-red-600">
                  ...more
                </button>
              )}
            </h1>

            <div className="flex items-center justify-end gap-1 py-2">
              <span className="cursor-pointer hover:text-slate-500">
                <AiTwotoneDelete />
              </span>
              <span className="cursor-pointer hover:text-slate-500">
                <AiTwotoneEdit />
              </span>
              <span className="cursor-pointer hover:text-slate-500">
                <AiOutlineCheckCircle />
              </span>
            </div>
          </div>

          {showText && (
            <div className="absolute inset-0 flex items-center justify-center bg-slate-600/75">
              <span
                onClick={() => setShowText(false)}
                className="absolute top-10 right-10 text-xl cursor-pointer hover:text-slate-500">
                <FaTimes />
              </span>
              <div className="w-[25rem] h-[25rem] bg-todo rounded-md">
                <Layer text={showFullText} />
              </div>
            </div>
          )}
        </div>
      ))}
    </div>
  );
};

export { TodoItem }
